ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java

```diff
@@ -1857,6 +1870,32 @@ FindInfo findEntriesSince(FindToken token, long maxTotalSizeOfEntries) throws St
     }
   }
 
+  /**
+   * Return true to stop scanning the index segment files. We stop scanning when
+   * 1. storeDeletedPutAsDeleteInFindEntries is false or
+   * 2. No message entries was found in the last scan or
+   * 3. new Token is not IndexBased anymore or
+   * 4. the total size of returned message infos are already larger than the max total size of entries.
```

Comment:
Does findEntriesFromSegmentStartOffset faithfully make sure the entry size is less or equal than required size?
If yes, so for the last loop, if the required size is very small like 10 bytes and findEntriesFromSegmentStartOffset will return nothing. But we are good since the shouldStopScanning will stop if messageEntries.isEmpty().
In this case, messageInfoSizeSum only can be equal to maxTotalSizeOfEntries but never exceed it, right?

Reply:
not really, the sum of message info size can be greater than the maxTotalSize. The way it works it 
```
for each indexvalue:
    if currentSize < max size: 
        add index value to the list
    else:
        break
    currentSize += index value.size
```
as you can see if the max size is 10, but the first index value's size is 100, we still add the first index value to the list, but we will not add the second one.

Reply:
not just double. Imagine you set the max total size to be 100 but one PutBlob is 4MB, we would still send this PutBlob back to requester.

Reply:
Yes, this is what I mean. Currently maxTotalSizeOfEntries is 4MB. We may send back 8MB entries.